### DISCUSSION TITLE
PFS on macOS

### DISCUSSION THREAD

**Discussion by Kentzo** [strongswan/strongswan-docs#16] [General]
It appears that [PFS is disabled by defaulton macOS](https://developer.apple.com/documentation/devicemanagement/vpn/ikev2) and must be explicitly enabled via the profile. The net result seems to be that non-profile configs will fail to rekey if esp_proposals specify a D-H group.

On the other hand, on iOS PFS seems to be enabled by default.

**Comment by tobiasbrunner** [comment]
> The net result seems to be that non-profile configs will fail to rekey if esp_proposals specify a D-H group.

Which is not currently recommended [in the docs](https://docs.strongswan.org/docs/5.9/interop/ios.html).

> On the other hand, on iOS PFS seems to be enabled by default.

How do you come to this conclusion? The docs say "Default: 0" for EnablePFS so doesn't that apply to iOS? What are the actual proposals sent during CHILD_SA rekeying by the different clients (and possibly macOS/iOS versions)?

**Comment by Kentzo** [comment]
I do not see this caveat explicitly mentioned there, hence this request.On iOS it seems to re-key successfully when esp proposals specify the d-h group. I will do more testing and report here. Also see https://developer.apple.com/forums/thread/659209

Btw, to collect IKEv2 logs on macOS one may use the “process:NEIKEv2Provider” search filter in Console.app, at least since Ventura.

**Reply by tobiasbrunner** [reply]
> I do not see this caveat explicitly mentioned there, hence this request.

Such a mismatch is a general issue with IKEv2 that's e.g. mentioned [here](https://docs.strongswan.org/docs/5.9/config/rekeying.html#_ikev2).

**Reply by Kentzo** [reply]
Please consider mentioning this caveat explicitly in the interoperability wiki. When I first faced this issue, I couldn’t immediately locate the problem and had to resort to tinkering.

**Comment by Kentzo** [comment]
strongSwan 5.7.2:

Phase 1:
- IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
- IKE:AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024

Phase 2:
- ESP:AES_CBC_256/HMAC_SHA2_256_128/MODP_2048/NO_EXT_SEQ
- ESP:AES_CBC_128/HMAC_SHA1_96/MODP_1024/NO_EXT_SEQ
- ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ
- ESP:AES_CBC_128/HMAC_SHA1_96/NO_EXT_SEQ

### iOS 16.5:

IKEv2 configured directly in the Settings, no other configuration profiles involved.

Phase 1:
- **IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048**
- IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_256
- IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1536
- IKE:AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
- IKE:3DES_CBC/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024

Phase 2:
- **ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ**
- ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ
- ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ
- ESP:AES_CBC_128/HMAC_SHA1_96/NO_EXT_SEQ
- ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ

Rekey (after ~8 minutes):
- **IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048**
- **ESP:AES_CBC_256/HMAC_SHA2_256_128/MODP_2048/NO_EXT_SEQ**
- ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ


### macOS 13.4:

IKEv2 configured directly in the Settings, no other configuration profiles involved.

Phase 1:
- **IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048**

Phase 2:
- **ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ**

Rekey (after ~24min):

- **ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ**

**Reply by tobiasbrunner** [reply]
> strongSwan 5.7.2:

What do you mean? These are definitely not the default proposals. Or did you list those just for reference regarding your setup?

> iOS 16.5:
> macOS 13.4:

Thanks for these. Interesting that iOS sends an ESP proposal with and one without DH group, while the macOS client doesn't do that. That it only sends a single proposal is quite strange (although the proposed algorithms are pretty much universally supported, I guess).

By the way, they could have simplified this by sending NONE as second DH transform in the first ESP proposal. Similar for the first three IKE proposals, which could be combined and all three DH groups listed in one proposal. That they send three duplicate ESP proposals is for the same reason (they have an ESP proposal for each DH group, but during IKE_AUTH the DH groups are stripped, resulting these duplicates).

I assume if the server selected the IKE proposal with _ecp256_, the first ESP proposal sent by iOS would use that DH group as well.

**Reply by Kentzo** [reply]
strongSwan proposals were given for reference. iOS and macOS are the proposals made by default. 

I noticed that you maintain per-version recommendation and wanted to share this so you could update to reflect more recent versions.

**Reply by tobiasbrunner** [reply]
I updated [the iOS/macOS docs](https://docs.strongswan.org/docs/5.9/interop/ios.html) earlier today.

**Reply by Kentzo** [reply]
Could you add the “ Troubleshooting IKEv**2** on macOS” section using my comment above too?

**Reply by tobiasbrunner** [reply]
Done!

**Reply by Kentzo** [reply]
FWIW, the CLI alternative is:
- `log show --predicate 'process == "NEIKEv2Provider"'` to show past log messages
- `log stream --predicate 'process == "NEIKEv2Provider"'` to stream in realtime

**Reply by tobiasbrunner** [reply]
Thanks, added that too.
